### PR TITLE
[DTensor] Change Sharding algorithm to be in line with ``torch.chunk()``

### PR DESCRIPTION
**Pull request by wz337**
As functional collective being updated, using tensor_split() as the underlying sharding algorithm would require padding and unpadding on multiple ranks. Therefore, we are changing the sharding algorithm to be in line with ``torch.chunk()`` to allow padding on the last two ranks in most of the scenarios.

### CODE REVIEW

File: torch/distributed/_tensor/placement_types.py [hunk 1]
```diff
@@ -54,37 +54,102 @@ def _split_tensor(
...
-    def _pad_tensor(self, tensor: torch.Tensor) -> torch.Tensor:
-        # pad tensor by 1 on the shard dim
-        pad = [0, 0] * (tensor.ndim - self.dim)
-        pad[-1] = 1
-        return torch.nn.functional.pad(tensor, pad)
+    def _pad_tensor(
+        self,
+        tensor: torch.Tensor,
+        pad_size: int,
+        reference_tensor: Optional[torch.Tensor] = None,
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:110]: I think we probably don't need this `reference_tensor` anymore if we creating the empty tensor with sth like `tensor.new_zeros((0, 3))`, we only need the pad_size and that would make the padding logic consistent too.

File: torch/distributed/_tensor/placement_types.py [hunk 2]
```diff
@@ -54,37 +54,102 @@ def _split_tensor(
...
+            device = torch.device(
+                tensor.get_device() if torch.cuda.is_available() else "cpu"
+            )
+            tensor_size = list(reference_tensor.size())
+            tensor_size = [dim if dim >= self.dim else 0 for dim in tensor_size]  # type: ignore[attr-defined]
+            return torch.zeros(tensor_size, device=device)
+        else:
+            pad = [0, 0] * (tensor.ndim - self.dim)
+            pad[-1] = pad_size
+            return torch.nn.functional.pad(tensor, pad)
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:122]: I think padding also works for tensor like `torch.tensor([], size=(0, 3))`

File: torch/distributed/_tensor/placement_types.py [hunk 3]
```diff
@@ -102,15 +167,27 @@ def _local_shard_size_on_dim(
...
+                min(size_on_dim, full_chunk_size * (idx + 1)) - full_chunk_size * idx,
+                0,
+            )
+            for idx in range(num_chunks)
+        ]
+        local_shard_size = chunk_sizes[rank]
+
         local_offset_on_dim = -1
         if return_offset:
+            # QQ: what would be the offset of an empty shard? -1?
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:186]: hmmm that's a good point... do you know how sharded tensor offset on empty shard looks like? I think we might can return the "global tensor dim size" for empty shard (representing the end of that tensor dim), but would like to see if this make sense for existing use case.

File: torch/distributed/_tensor/placement_types.py [hunk 4]
```diff
@@ -39,7 +39,7 @@ def _split_tensor(
         *,
         with_padding: bool = True,
         contiguous: bool = True,
-    ) -> Tuple[List[torch.Tensor], int]:
+    ) -> Tuple[List[torch.Tensor], int, List[int]]:
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:42]: so it looks like all the callsites (except `test_device_mesh.py` don't need the second return argument anymore (i think it's because this embeds in the last return argument). Shall we delete the second return arg and return two args `Tuple[List[torch.Tensor], List[int]]` instead?

File: torch/distributed/_tensor/placement_types.py [hunk 5]
```diff
@@ -54,37 +54,102 @@ def _split_tensor(
...
+        # compute the chunk size inline with ``torch.chunk``
+        full_chunk_size = (tensor.size(self.dim) + num_chunks - 1) // num_chunks
+
+        # Compute chunk size for each chunk for ``self.dim``
+        chunk_sizes = [
+            tensor_list[idx].size(self.dim) if idx < len(tensor_list) else 0
+            for idx in range(num_chunks)
+        ]
+        # Get idx start to pad
+        idx_start_to_pad = next(
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:69]: do we really need this given that we only need `pad_sizes`? `pad_sizes` can actually be computed without this iiuc.

File: torch/distributed/_tensor/placement_types.py [hunk 6]
```diff
@@ -54,37 +54,102 @@ def _split_tensor(
...
+                idx
+                for idx, _ in enumerate(chunk_sizes)
+                if idx > 0 and chunk_sizes[idx] < chunk_sizes[idx - 1]
+            ),
+            0,
+        )
+        # Compute pad size on each chunk
+        pad_sizes = [
+            full_chunk_size - chunk_size if idx >= idx_start_to_pad else 0
+            for idx, chunk_size in enumerate(chunk_sizes)
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:80]: i.e. we don't really need to check `if idx >=idx_start_to_pad`? it could always be `full_chunk_size-chunk_size`, for ranks don't need to pad, the subtraction would become 0 automatically?

**Review by wanchaol — COMMENTED** [review]
First pass, looks pretty good already, I have a few suggestions inlined. Mainly I think we should use `new_zeros` to create empty shard, which could help us consolidate the padding logic and no need to infer the device :)

File: torch/distributed/_tensor/placement_types.py (hunk 4 above)
Comment (wz337) [reply torch/distributed/_tensor/placement_types.py:42]: Good point! Actually thinking about the same thing!

File: torch/distributed/_tensor/placement_types.py (hunk 3 above)
Comment (wz337) [reply torch/distributed/_tensor/placement_types.py:186]: Let me double check on this.

File: torch/distributed/_tensor/placement_types.py [hunk 7]
```diff
@@ -54,40 +54,70 @@ def _split_tensor(
...
+        full_chunk_size = (tensor.size(self.dim) + num_chunks - 1) // num_chunks
+
+        # Compute chunk size for each chunk for ``self.dim``
+        chunk_sizes = [
+            tensor_list[idx].size(self.dim) if idx < len(tensor_list) else 0
+            for idx in range(num_chunks)
+        ]
+        # Compute pad size on each chunk
+        pad_sizes = [
+            full_chunk_size - chunk_size for idx, chunk_size in enumerate(chunk_sizes)
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:70]: nit: no need to have `idx` here?

File: torch/distributed/_tensor/placement_types.py [hunk 8]
```diff
@@ -182,27 +227,44 @@ def _to_replicate_tensor(
...
+        chunk_sizes = [
+            max(
+                min(size[self.dim], full_chunk_size * (idx + 1))
+                - full_chunk_size * idx,
+                0,
+            )
+            for idx in range(num_chunks)
+        ]
+        pad_sizes = [full_chunk_size - chunk_size for chunk_size in chunk_sizes]
+        is_padded = not all(pad_size == 0 for pad_size in pad_sizes)
```
Comment (wanchaol) [inline torch/distributed/_tensor/placement_types.py:247]: nit: why don't we use a similar condition for `is_padded` in `reduce_scatter_tensor`? i.e. `size[self.dim]% num_chunks != 0`

**Review by wanchaol — APPROVED** [review]
Looks great! Thanks for working on it! I have a few more suggestions inlined.

File: torch/distributed/_tensor/placement_types.py [hunk 9]
```diff
@@ -54,40 +54,70 @@ def _split_tensor(
...
+        tensor = tensor.narrow(
+            self.dim,
+            start=0,
+            length=tensor.size(self.dim) - pad_size,
+        )
+
+        # Explicitly return an empty tensor. Otherwise, even if the
+        # tensor is empty, the size won't be 0.
+        if tensor.numel() == 0:
+            return tensor.new_zeros([0])
```
[the remark this answers is not in the document]
Comment (wz337) [reply torch/distributed/_tensor/placement_types.py:118]: > Hmmm could you explain why we need to explicitly return an empty tensor? iiuc after `tensor.narrow` there would be some ranks have `tensor` with shape ([0, 8]) for example, that's still considered as empty tensor with no data, so that should work for us.

The reason is that we have test to compare an unpad tensor with original tensor_to_split.

https://github.com/pytorch/pytorch/blob/ada67c9d8a3ac61fa0af5b8a186d5ecb31765af5/test/distributed/_tensor/test_device_mesh.py#L211

In this case, the size of the two would fail the assert as one could be ([0, 8]) and the other one is ([0]), although both are just empty tensors([]). I guess I could update the test cases to make _unpad() more consistent.
